File: apps/android/app/src/main/java/pro/wewed/app/services/RoleScopedAccess.kt
package pro.wewed.app.services

import kotlinx.coroutines.sync.Mutex
import kotlinx.coroutines.sync.withLock
import pro.wewed.app.models.*

/** Admin support reads are recorded before any data is returned. */
class AdminAuditLog {
    private val mutex = Mutex()
    private val entries = mutableListOf<AdminAuditEntry>()

    suspend fun record(section: String, weddingId: String, nowMillis: Long = System.currentTimeMillis()) = mutex.withLock {
        entries.add(0, AdminAuditEntry(section, weddingId, nowMillis))
    }

    suspend fun all(): List<AdminAuditEntry> = mutex.withLock { entries.toList() }
}

/**
 * The only data gateway a role shell receives. Each call checks the grant's capabilities and scope,
 * so a guest, vendor or usher shell cannot read the roster, budget or financials even if a screen tried.
 */
class RoleScopedAccess(
    val grant: RoleGrant,
    private val wedding: WeddingRepository,
    private val planner: PlannerDashboardRepository,
    private val audit: AdminAuditLog = AdminAuditLog()
) {
    val role: AppRole get() = grant.role

    fun can(capability: Capability): Boolean = CapabilityPolicy.allows(grant.role, capability)

    private fun require(capability: Capability) {
        if (!can(capability)) throw AccessDeniedException(grant.role, capability)
    }

    // Public wedding information

    suspend fun weddingSummary(): Wedding {
        require(Capability.VIEW_WEDDING_SUMMARY)
        return wedding.getWedding()
    }

    suspend fun programme(): List<PlannerTimelineEntry> {
        require(Capability.VIEW_PROGRAMME)
        return planner.getTimelineEntries()
    }

    suspend fun announcements(): List<WeddingAnnouncement> {
        require(Capability.VIEW_WEDDING_SUMMARY)
        return wedding.getAnnouncements()
    }

    // Wedding management (Couple / Planner)

    suspend fun dashboard(): PlannerDashboardSnapshot {
        require(Capability.VIEW_PLANNING_DASHBOARD)
        return planner.getDashboard()
    }

    suspend fun tasks(): List<PlannerTask> {
        require(Capability.VIEW_TASKS)
        return wedding.getTasks()
    }

    suspend fun createTask(title: String, priority: TaskPriority, category: String): PlannerTask {
        require(Capability.MANAGE_TASKS)
        return wedding.createTask(title, priority, category)
    }

    suspend fun toggleTask(taskId: String): PlannerTask {
        require(Capability.MANAGE_TASKS)
        return wedding.toggleTask(taskId)
    }

    suspend fun guestRoster(): List<Guest> {
        require(Capability.VIEW_GUEST_ROSTER)
        return wedding.getGuests()
    }

    suspend fun searchGuests(query: String): List<Guest> {
        require(Capability.VIEW_GUEST_ROSTER)
        return wedding.searchGuests(query)
    }

    suspend fun budget(): BudgetSummary {
        require(Capability.VIEW_BUDGET)
        return wedding.getBudget()
    }

    suspend fun budgetLines(): List<PlannerBudgetLine> {
        require(Capability.VIEW_BUDGET)
        return planner.getBudgetLines()
    }

    suspend fun contributions(): List<PlannerContributionRecord> {
        require(Capability.VIEW_CONTRIBUTIONS)
        val showIdentity = can(Capability.VIEW_CONTRIBUTOR_IDENTITY)
        return planner.getContributions().map { record -> redactContributor(record, showIdentity) }
    }

    suspend fun vendorEngagements(): List<PlannerVendorEngagement> {
        val all = planner.getVendorEngagements()
        return when {
            can(Capability.VIEW_ALL_VENDOR_ENGAGEMENTS) -> all
            can(Capability.VIEW_OWN_VENDOR_ENGAGEMENT) -> {
                val vendorId = grant.vendorId ?: return emptyList()
                all.filter { it.vendorId == vendorId }
            }
            else -> throw AccessDeniedException(grant.role, Capability.VIEW_OWN_VENDOR_ENGAGEMENT)
        }
    }

    suspend fun vendorPresence(): List<VendorPresence> {
        val all = wedding.getVendors()
        return when {
            can(Capability.VIEW_VENDOR_PRESENCE) -> all
            can(Capability.UPDATE_OWN_VENDOR_PRESENCE) -> {
                val vendorId = grant.vendorId ?: return emptyList()
                all.filter { it.id == vendorId }
            }
            else -> throw AccessDeniedException(grant.role, Capability.VIEW_VENDOR_PRESENCE)
        }
    }

    suspend fun updateOwnVendorPresence(state: VendorPresenceState): VendorPresence {
        require(Capability.UPDATE_OWN_VENDOR_PRESENCE)
        val vendorId = grant.vendorId ?: throw AccessDeniedException(grant.role, Capability.UPDATE_OWN_VENDOR_PRESENCE)
        return wedding.updateVendorState(vendorId, state)
    }

    suspend fun seating(): List<PlannerSeatingTable> {
        require(Capability.VIEW_SEATING)
        return planner.getSeatingTables()
    }

    suspend fun documents(): List<PlannerDocumentRecord> {
        require(Capability.VIEW_DOCUMENTS)
        return planner.getDocuments()
    }

    suspend fun updateWeddingDetails(update: WeddingDetailsUpdate): Wedding {
        require(Capability.EDIT_WEDDING_DETAILS)
        return wedding.updateWeddingDetails(update)
    }

    suspend fun previewGuestPass(token: String): WeddingPass {
        require(Capability.PREVIEW_GUEST_PASSES)
        return wedding.getWeddingPass(token)
    }

    suspend fun postAnnouncement(title: String, message: String, urgency: AnnouncementUrgency): WeddingAnnouncement {
        require(Capability.POST_ANNOUNCEMENT)
        return wedding.postAnnouncement(title, message, urgency)
    }

    // Guest: own invitation, RSVP and pass only

    private fun ownGuestToken(capability: Capability): String {
        require(capability)
        return grant.guestId ?: throw AccessDeniedException(grant.role, capability)
    }

    suspend fun ownInvitation(): InvitationContext {
        val token = ownGuestToken(Capability.VIEW_OWN_INVITATION)
        return wedding.resolveInvitation(grant.weddingId, token)
    }

    suspend fun respondToOwnInvitation(attending: Boolean): WeddingPass {
        val token = ownGuestToken(Capability.RESPOND_OWN_RSVP)
        return wedding.confirmRsvp(grant.weddingId, token, attending)
    }

    /** Null until the guest has accepted; a declined or pending guest has no admission pass. */
    suspend fun ownPass(): WeddingPass? {
        val token = ownGuestToken(Capability.VIEW_OWN_PASS)
        val invitation = wedding.resolveInvitation(grant.weddingId, token)
        if (!invitation.isConfirmed) return null
        return wedding.getWeddingPass(token)
    }

    // Gate operations (Usher / Coordinator / Couple / Planner)

    suspend fun admit(qrPayload: String, count: Int, operatorId: String): CheckInVerificationResult {
        require(Capability.SCAN_ADMISSION)
        return wedding.checkInGuest(qrPayload, count, operatorId)
    }

    suspend fun admissionLookup(query: String): List<AdmissionLookupRow> {
        require(Capability.LOOKUP_ADMISSION)
        val normalized = query.trim().lowercase()
        if (normalized.length < 2) return emptyList()
        return wedding.getGuests()
            .filter { it.rsvpStatus == RSVPStatus.ATTENDING && it.name.lowercase().contains(normalized) }
            .map { AdmissionLookupRow(it.id, it.name, it.partySize, it.checkedInCount, it.tableName) }
    }

    suspend fun admissionSummary(): AdmissionSummary {
        require(Capability.VIEW_ADMISSION_SUMMARY)
        val attending = wedding.getGuests().filter { it.rsvpStatus == RSVPStatus.ATTENDING }
        return AdmissionSummary(
            attendingParties = attending.size,
            expectedGuests = attending.sumOf { it.partySize },
            admittedGuests = attending.sumOf { it.checkedInCount }
        )
    }

    suspend fun admissionHistory(): List<CheckInAuditRecord> {
        require(Capability.SCAN_ADMISSION)
        return wedding.getAuditRecords()
    }

    // Admin: explicit, audited support reads

    suspend fun <T> supportRead(
        section: String,
        block: suspend (WeddingRepository, PlannerDashboardRepository) -> T
    ): T {
        require(Capability.ADMIN_SUPPORT)
        audit.record(section, grant.weddingId)
        return block(wedding, planner)
    }

    suspend fun auditEntries(): List<AdminAuditEntry> {
        require(Capability.ADMIN_SUPPORT)
        return audit.all()
    }

    companion object {
        fun redactContributor(record: PlannerContributionRecord, roleMayIdentify: Boolean): PlannerContributionRecord {
            if (record.contributorResolution == ContributorResolution.NOT_RECORDED) return record
            val anonymous = record.privacyLabel.equals("Anonymous", ignoreCase = true)
            return if (roleMayIdentify && !anonymous) record else record.copy(
                contributorLabel = if (anonymous) "Anonymous" else "Contributor hidden",
                contributorGuestId = null,
                contributorResolution = ContributorResolution.HIDDEN
            )
        }
    }
}

File: apps/ios/Wewed/Services/RoleScopedAccess.swift
import Foundation

/// Admin support reads are recorded before any data is returned.
public actor AdminAuditLog {
    private var entries: [AdminAuditEntry] = []

    public init() {}

    public func record(section: String, weddingId: String, at date: Date = Date()) {
        entries.insert(AdminAuditEntry(section: section, weddingId: weddingId, recordedAt: date), at: 0)
    }

    public func all() -> [AdminAuditEntry] { entries }
}

/// The only data gateway a role shell receives. Each call checks the grant's capabilities and scope,
/// so a guest, vendor or usher shell cannot read the roster, budget or financials even if a screen tried.
/// Mirrors Android `RoleScopedAccess`.
public final class RoleScopedAccess: @unchecked Sendable {
    public let grant: RoleGrant
    private let wedding: WeddingRepositoryProtocol
    private let planner: PlannerDashboardRepositoryProtocol
    private let audit: AdminAuditLog

    public init(
        grant: RoleGrant,
        wedding: WeddingRepositoryProtocol,
        planner: PlannerDashboardRepositoryProtocol,
        audit: AdminAuditLog = AdminAuditLog()
    ) {
        self.grant = grant
        self.wedding = wedding
        self.planner = planner
        self.audit = audit
    }

    public var role: AppRole { grant.role }

    public func can(_ capability: Capability) -> Bool {
        CapabilityPolicy.allows(grant.role, capability)
    }

    private func require(_ capability: Capability) throws {
        guard can(capability) else { throw AccessDeniedError(role: grant.role, capability: capability) }
    }

    // MARK: Public wedding information

    public func weddingSummary() async throws -> Wedding {
        try require(.viewWeddingSummary)
        return try await wedding.getWedding()
    }

    public func programme() async throws -> [PlannerTimelineEntry] {
        try require(.viewProgramme)
        return try await planner.getTimelineEntries()
    }

    public func announcements() async throws -> [WeddingAnnouncement] {
        try require(.viewWeddingSummary)
        return try await wedding.getAnnouncements()
    }

    // MARK: Wedding management (Couple / Planner)

    public func dashboard() async throws -> PlannerDashboardSnapshot {
        try require(.viewPlanningDashboard)
        return try await planner.getDashboard()
    }

    public func tasks() async throws -> [PlannerTask] {
        try require(.viewTasks)
        return try await wedding.getTasks()
    }

    public func createTask(title: String, priority: TaskPriority, category: String) async throws -> PlannerTask {
        try require(.manageTasks)
        return try await wedding.createTask(title: title, priority: priority, category: category)
    }

    public func toggleTask(taskId: String) async throws -> PlannerTask {
        try require(.manageTasks)
        return try await wedding.toggleTask(taskId: taskId)
    }

    public func guestRoster() async throws -> [Guest] {
        try require(.viewGuestRoster)
        return try await wedding.getGuests()
    }

    public func searchGuests(query: String) async throws -> [Guest] {
        try require(.viewGuestRoster)
        return try await wedding.searchGuests(query: query)
    }

    public func budget() async throws -> BudgetSummary {
        try require(.viewBudget)
        return try await wedding.getBudget()
    }

    public func budgetLines() async throws -> [PlannerBudgetLine] {
        try require(.viewBudget)
        return try await planner.getBudgetLines()
    }

    public func contributions() async throws -> [PlannerContributionRecord] {
        try require(.viewContributions)
        let showIdentity = can(.viewContributorIdentity)
        return try await planner.getContributions().map { Self.redactContributor($0, roleMayIdentify: showIdentity) }
    }

    public func vendorEngagements() async throws -> [PlannerVendorEngagement] {
        let all = try await planner.getVendorEngagements()
        if can(.viewAllVendorEngagements) { return all }
        if can(.viewOwnVendorEngagement) {
            guard let vendorId = grant.vendorId else { return [] }
            return all.filter { $0.vendorId == vendorId }
        }
        throw AccessDeniedError(role: grant.role, capability: .viewOwnVendorEngagement)
    }

    public func vendorPresence() async throws -> [VendorPresence] {
        let all = try await wedding.getVendors()
        if can(.viewVendorPresence) { return all }
        if can(.updateOwnVendorPresence) {
            guard let vendorId = grant.vendorId else { return [] }
            return all.filter { $0.id == vendorId }
        }
        throw AccessDeniedError(role: grant.role, capability: .viewVendorPresence)
    }

    public func updateOwnVendorPresence(_ state: VendorPresenceState) async throws -> VendorPresence {
        try require(.updateOwnVendorPresence)
        guard let vendorId = grant.vendorId else {
            throw AccessDeniedError(role: grant.role, capability: .updateOwnVendorPresence)
        }
        return try await wedding.updateVendorState(id: vendorId, state: state)
    }

    public func seating() async throws -> [PlannerSeatingTable] {
        try require(.viewSeating)
        return try await planner.getSeatingTables()
    }

    public func documents() async throws -> [PlannerDocumentRecord] {
        try require(.viewDocuments)
        return try await planner.getDocuments()
    }

    public func updateWeddingDetails(_ update: WeddingDetailsUpdate) async throws -> Wedding {
        try require(.editWeddingDetails)
        return try await wedding.updateWeddingDetails(update)
    }

    public func previewGuestPass(token: String) async throws -> WeddingPass {
        try require(.previewGuestPasses)
        return try await wedding.getWeddingPass(token: token)
    }

    public func postAnnouncement(title: String, message: String, urgency: AnnouncementUrgency) async throws -> WeddingAnnouncement {
        try require(.postAnnouncement)
        return try await wedding.postAnnouncement(title: title, message: message, urgency: urgency)
    }

    // MARK: Guest: own invitation, RSVP and pass only

    private func ownGuestToken(_ capability: Capability) throws -> String {
        try require(capability)
        guard let guestId = grant.guestId else { throw AccessDeniedError(role: grant.role, capability: capability) }
        return guestId
    }

    public func ownInvitation() async throws -> InvitationContext {
        let token = try ownGuestToken(.viewOwnInvitation)
        return try await wedding.resolveInvitation(weddingSlug: grant.weddingId, token: token)
    }

    public func respondToOwnInvitation(attending: Bool) async throws -> WeddingPass {
        let token = try ownGuestToken(.respondOwnRsvp)
        return try await wedding.confirmRsvp(weddingSlug: grant.weddingId, token: token, attending: attending)
    }

    /// Nil until the guest has accepted; a declined or pending guest has no admission pass.
    public func ownPass() async throws -> WeddingPass? {
        let token = try ownGuestToken(.viewOwnPass)
        let invitation = try await wedding.resolveInvitation(weddingSlug: grant.weddingId, token: token)
        guard invitation.isConfirmed else { return nil }
        return try await wedding.getWeddingPass(token: token)
    }

    // MARK: Gate operations (Usher / Coordinator / Couple / Planner)

    public func admit(qrPayload: String, count: Int, operatorId: String) async throws -> CheckInVerificationResult {
        try require(.scanAdmission)
        return try await wedding.checkInGuest(qrPayload: qrPayload, count: count, usherId: operatorId)
    }

    public func admissionLookup(query: String) async throws -> [AdmissionLookupRow] {
        try require(.lookupAdmission)
        let normalized = query.trimmingCharacters(in: .whitespacesAndNewlines).lowercased()
        guard normalized.count >= 2 else { return [] }
        return try await wedding.getGuests()
            .filter { $0.rsvpStatus == .attending && $0.name.lowercased().contains(normalized) }
            .map {
                AdmissionLookupRow(
                    guestId: $0.id,
                    displayName: $0.name,
                    partySize: $0.partySize,
                    admittedCount: $0.checkedInCount,
                    tableName: $0.tableName
                )
            }
    }

    public func admissionSummary() async throws -> AdmissionSummary {
        try require(.viewAdmissionSummary)
        let attending = try await wedding.getGuests().filter { $0.rsvpStatus == .attending }
        return AdmissionSummary(
            attendingParties: attending.count,
            expectedGuests: attending.reduce(0) { $0 + $1.partySize },
            admittedGuests: attending.reduce(0) { $0 + $1.checkedInCount }
        )
    }

    public func admissionHistory() async throws -> [CheckInAuditRecord] {
        try require(.scanAdmission)
        return try await wedding.getAuditRecords()
    }

    // MARK: Admin: explicit, audited support reads

    public func supportRead<T>(
        section: String,
        _ block: (WeddingRepositoryProtocol, PlannerDashboardRepositoryProtocol) async throws -> T
    ) async throws -> T {
        try require(.adminSupport)
        await audit.record(section: section, weddingId: grant.weddingId)
        return try await block(wedding, planner)
    }

    public func auditEntries() async throws -> [AdminAuditEntry] {
        try require(.adminSupport)
        return await audit.all()
    }

    public static func redactContributor(_ record: PlannerContributionRecord, roleMayIdentify: Bool) -> PlannerContributionRecord {
        if record.contributorResolution == .notRecorded { return record }
        let anonymous = record.privacyLabel?.caseInsensitiveCompare("Anonymous") == .orderedSame
        if roleMayIdentify && !anonymous { return record }
        return PlannerContributionRecord(
            id: record.id,
            contributorLabel: anonymous ? "Anonymous" : "Contributor hidden",
            typeLabel: record.typeLabel,
            value: record.value,
            statusLabel: record.statusLabel,
            allocationLabel: record.allocationLabel,
            verified: record.verified,
            contributorGuestId: nil,
            contributorResolution: .hidden,
            privacyLabel: record.privacyLabel,
            wordCount: record.wordCount,
            submittedAtLabel: record.submittedAtLabel,
            messageText: record.messageText
        )
    }
}
